CARDS-1741: Add a "PageStart" extension point to the patient UI

veronikaslc opened the pull request:
Also fixes [CARDS-1929](https://phenotips.atlassian.net/browse/CARDS-1929), [CARDS-1930](https://phenotips.atlassian.net/browse/CARDS-1930), [CARDS-1931](https://phenotips.atlassian.net/browse/CARDS-1931).

To test:
- Start yourexperience or cards in test mode, with --demo to activate the demo banner
- From Administration, enable the Downtime Warning to have two stacked banners
- Create a visit with surveys, then obtain the survey link as usual
- As patient, using the survey link, check that the banners are displayed as expected in the patient portal

This is a relatively LOW_RISK pr because of the reduced visibility of PageStart banners.

[CARDS-1929]: https://phenotips.atlassian.net/browse/CARDS-1929?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CARDS-1930]: https://phenotips.atlassian.net/browse/CARDS-1930?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CARDS-1931]: https://phenotips.atlassian.net/browse/CARDS-1931?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

veronikaslc committed:
CARDS-1741: Add a PageStart extension point to the patient UI

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/index.jsx:
```diff
@@ -111,16 +112,21 @@ function PatientPortalHomepage (props) {
-      </Router>
-    </ThemeProvider>
-  </StyledEngineProvider>,
+  <React.Fragment>
```
Why do we need a React.Fragment? AFAICS there's only one child.

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/index.jsx:
```diff
@@ -111,16 +112,21 @@ function PatientPortalHomepage (props) {
+  <React.Fragment>
+    <StyledEngineProvider injectFirst>
+      <ThemeProvider theme={appTheme}>
+        <PageStart
+      extensionsName="SurveyPageStart"
+    />
```
Messy indentation

sashaandjic commented:
The banner is not visible when patient logs in.
![image](https://user-images.githubusercontent.com/23243244/191324350-7cc06606-6455-4144-81ea-e46fcb8d0405.png)

Also, should we add banner here too:
![image](https://user-images.githubusercontent.com/23243244/191324450-f30eed75-e66e-4d5b-aa3f-cca38c79d192.png)

sashaandjic commented:
Steps to reproduce:
- start cards in either `proms` or `prems` mode
- login as admin
- create a Patient Info and Visit Info forms --> observe the surveys that are created
- logout and login as patient
- when survey form are opened notice that the banner is present but almost completely covered - only small slices on each end are visible

Please note: if in `prems` mode, you would need to generate a token to login as patient, as patients can only use links (with token) that are going to be mailed to them.

marta- reviewed (changes requested):
A couple of layout issues popped up in the Patient UI with the introduction of the Demo banner, and they need to be addressed:
* On a short screen (e.g. on a smartphone or a short desktop browser window) the footer containing the link to the Terms of Use is pushed too high and ends up being displayed above the action button. It should be positioned at the bottom, like it is when there is no banner.
![image](https://user-images.githubusercontent.com/651980/191617981-44ab8c40-aa0e-4d12-84ae-2984846658ca.png)


* After logging in, when scrolling a survey up, the banner is visible on the sides behind the survey header. It should not be visible at all once the user starts scrolling up.
![image](https://user-images.githubusercontent.com/651980/191618254-e625ddc8-b23c-4279-a481-3f54eb72f582.png)
![image](https://user-images.githubusercontent.com/651980/191618492-cb58db1b-d569-4d87-8c0c-a83dcc17a449.png)
* The pagination controls look misplaced when the banner is on:
![image](https://user-images.githubusercontent.com/651980/191618723-d47c2459-4ab6-4163-8a12-44574ba60b4e.png)

To replicate all of the above:
* start with `-P proms --demo`
* as admin, create patient info form and visit info form; the latter should have `6012-HC-Congenital Cardiac` selected for the `Clinic` question to make sure there's a paginated survey to fill out
* from Chrome dev tools, switch to emulating a mobile device (the above screenshots are on emulated iPhone SE), then log in and complete surveys as a patient.

marta- reviewed (changes requested):
1. The absolute positioning of the banners does not work well in the default UI (admin and clinician), as the banners get scrolled up leaving an unsightly gap at the top (see screenshot).
![image](https://user-images.githubusercontent.com/651980/191766789-e70b03bf-97e1-4155-9620-107a600673be.png)
I suggest passing the position as a prop so we can customize it on a case by case basis.
2. All looks good now in the patient UI with one minor exception: the shadow of the banner visible on the sides of the patient ui header looks a bit messy. I suggest we get rid of that shadow from all banners.
![image](https://user-images.githubusercontent.com/651980/191768839-8cd8bdfd-da0e-4512-a0ee-8bb3347d2a12.png)
3. There is still a screen without a banner:
  * In Administration/Patient identification, uncheck the first checkbox ("Patients can answer surveys without a personalized link")
  * Sign out and go to /Survey.html: a page displaying the welcome message and informing the user that "**To fill out surveys, please follow the personalized link that was emailed to you.**" is displayed. There is no banner on that page.
![image](https://user-images.githubusercontent.com/651980/191770252-1c387ea6-02f8-47c5-9f94-d86b705d46c2.png)

sashaandjic commented:
no banner here too:
![image](https://user-images.githubusercontent.com/23243244/191825442-014cff0a-a445-471c-b150-459e2852ce28.png)

sashaandjic commented:
Cannot access Admin button on narrow screens (probably out of scope)
![image](https://user-images.githubusercontent.com/23243244/191825639-3a990edb-1eb8-46de-9dbd-8baa63d62116.png)

sashaandjic commented:
Login page looks weird with both demo and maintenance banners:
![image](https://user-images.githubusercontent.com/23243244/191825867-668d0b25-e605-4328-881c-21f6b68c35f7.png)

marta- commented:
> Cannot access Admin button on narrow screens (probably out of scope)

Indeed it's out of scope.

> Login page looks weird with both demo and maintenance banners:

This one is out of scope too, since the login page extension point was developed elsewhere.

marta- commented:
> no banner here too:

The banner is there but it is under the landing page UI. TBH with the current implementation of the landing page it won't be trivial to display it. Additionally, given that the production workflows now exclude staff interacting with the app, this can be considered very low priority and should not condition the merge of this PR. We may even end up redoing the landing page completely, and we'll tackle the banner display then (or after).

veronikaslc commented:
All should be fixed except reverting to the `fixed` position did not seem to fix #1 `unsightly gap at the top` propblem.

sashaandjic commented:
small ui issue -not sure if it is related to this PR
Uploading Patient 1 _ DATA-PRO - Google Chrome 2022-09-23 09-40-27.mp4…

sashaandjic commented:
On iPad and cell phones patent cannot access the `BEGIN` button.

https://user-images.githubusercontent.com/23243244/191975677-f2d0f582-ec3b-4805-8f2b-601747eb330e.mp4

https://user-images.githubusercontent.com/23243244/191975695-37ef7d6b-5f5e-4d01-9db5-77acde6cac88.mp4

marta- commented:
> small ui issue -not sure if it is related to this PR

It is related to this PR and I wouldn't call it small. The introduction of `Collapse` [here](https://github.com/data-team-uhn/cards/pull/1180/files#diff-9d4d096ae6d13002709812b5e8fbf8369e1b6c087e123141a3e127fb2f399822R114) appears to be the cause.

marta- commented:
Also, in the patient ui identification page and start screen, the content is pushed down way too much:
![image](https://user-images.githubusercontent.com/651980/191978578-8ea61ca9-fd06-4488-951b-1c80bb8b869d.png)

![image](https://user-images.githubusercontent.com/651980/191978940-3ac0d1bd-4209-4002-8b08-c156fa0942a3.png)

veronikaslc commented:
Are you testing with 100% zoom or less?

veronikaslc commented:
Should collapse stay or it's better to resolve hiding with positioning?

marta- commented:
> Are you testing with 100% zoom or less?

100% zoom. It looks worse the more I zoom out.

marta- commented:
> Should collapse stay or it's better to resolve hiding with positioning?

In the current state it looks like Collapse might do more harm than good and we might want to look for a simpler option, if possible.

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/index.jsx:
```diff
@@ -86,25 +88,60 @@ function PatientPortalHomepage (props) {
+          messageColor="textPrimary"
+        />
+      </div>
+    </>)
```
This structure is repeated 3 times:
```
    <>
      <PageStart ... />
      <div style={ { position: 'relative', top: contentOffset + 'px' } }>
        { /* the content */ }
      </div>
    </>
```
It should be pulled either in a function or a separate component.

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/index.jsx:
```diff
@@ -86,25 +88,60 @@ function PatientPortalHomepage (props) {
-      <Footer />
+      <PageStart
+        extensionsName="SurveyPageStart"
+        zIndex="1300"
```
`zIndex` is a very low-level concept and shouldn't be passed as a property to a React component. Can the positioning be handled via a more generic prop?

veronikaslc replied:
Not sure I understand the comment, but in any case I will remove this fix as it was ment to fix the low priority landing page banner absence.

sashaandjic commented:
Something stick out from the background and prevent users from editing
![image](https://user-images.githubusercontent.com/23243244/192345217-f3a8c40b-72dd-47cc-a8ff-13096d0f4c69.png)

veronikaslc commented:
>>Can not recreate the above 
Failing to load `DashboardViews` is most likely out of the scope as this PR does not seem to alter extension loading or DashboardViews in particular. wdt @marta- 

>>Something stick out from the background and prevent users from editing
By some reason can not reproduce that one on my build

marta- commented:
> Failing to load `DashboardViews` is most likely out of the scope as this PR does not seem to alter extension loading or DashboardViews in particular. wdt @marta-

Indeed that's a poorly phrased log message. The cause of it is harmless, we simply don't have custom widgets displayed on the clinic dashboard (and we're not supposed to have them). Please ignore it. 

> 
> > Something stick out from the background and prevent users from editing
> 
> By some reason can not reproduce that one on my build

I can't replicate it either, but something did make the ghost element show up there at the bottom in that screenshot so I'll keep investigating.

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/192579604-bd84e2a2-e3b7-47a0-9a42-ef3c0418dee8.png)

veronikaslc commented:
![image](https://user-images.githubusercontent.com/4371097/192580362-ca1eaa8a-f0cc-46b4-b24f-054c1c14bc34.png)

sashaandjic commented:
Can this be improved?

https://user-images.githubusercontent.com/23243244/192580407-93b67997-3e5a-4eb3-ae9b-5f27b02dce49.mp4

sashaandjic commented:
@veronikaslc  this one is only happening in this PR
![image](https://user-images.githubusercontent.com/23243244/192581160-08e897db-f223-41f4-928f-6a2f313c9d91.png)

veronikaslc commented:
> Can this be improved?
> 
>  DATA-PRO.-.Google.Chrome.2022-09-27.12-17-16.mp4

This should be fixed now.

sashaandjic commented:
All reported defects except this one are now fixed:
https://user-images.githubusercontent.com/23243244/192674977-10ffc256-b634-4467-ba02-dc08ce7e8eab.mp4

Jira tasks [CARDS-1930](https://phenotips.atlassian.net/browse/CARDS-1930) and [CARDS-1931](https://phenotips.atlassian.net/browse/CARDS-1931) are fixed too.

veronikaslc commented:
> > no banner here too:
> 
> The banner is there but it is under the landing page UI. TBH with the current implementation of the landing page it won't be trivial to display it. Additionally, given that the production workflows now exclude staff interacting with the app, this can be considered very low priority and should not condition the merge of this PR. We may even end up redoing the landing page completely, and we'll tackle the banner display then (or after).

The only reply regarding landing page was that one.

sashaandjic reviewed (approved):
tested

The remaining issue is low priority and is reported in [CARDS-1934](https://phenotips.atlassian.net/browse/CARDS-1934)

veronikaslc committed:
CARDS-1741: Add a "PageStart" extension point to the patient UI

Mirror the existing PageStart / LoginPageStart extension points in the patient portal
so that content such as warning banners can be shown at the top of the survey pages.

veronikaslc committed:
CARDS-1929: Add a vertical scroll bar on very narrow screens

On very narrow (mobile) screens the page content could overflow instead of scrolling.
Constrain the heights so the content scrolls correctly.

- Limit the height of the sidebar wrapper so its content scrolls instead of running past
  the bottom of the viewport. The height also accounts for the PageStart banner offset.
- On narrow screens, stick the resource header to the top of the viewport instead of
  offsetting it by the banner height, because the banners do not push the content down at
  that size.

veronikaslc committed:
CARDS-1930: Keep the logo visible on the login page with two banners

Increase the top margin of the self-contained login box so the box and the logo
stay visible below up to two stacked LoginPageStart banners. The value is fixed
because the login page renders PageStart without the offset wrapper used on the
other pages.

veronikaslc committed:
CARDS-1931: Do not show the Downtime Warning Banner when dates are missing

If the configured from/to dates are missing or cannot be parsed, do not set them,
so the banner is not displayed at all (the component already renders nothing when
either date is unset). Use an isNaN check on the parsed date instead of comparing
the Date object to the "Invalid Date" string.

marta- committed:
CARDS-1741: Drop the unnecessary responsive positioning in the patient portal PageStart

PageStartWrapper and the patient portal Header were switching their positioning based on
screen size (an "appbarExpanded" media query), which seems  unnecessary.

Now the wrapped content is always positioned relative with a top offset equal to the banner
height, and the header always sticks just below the banners. This matches the approach
already implemented in the default UI.

marta- committed:
CARDS-1741: Fix banner overlap when the patient portal shows two PageStart banners

The Downtime Warning Banner renders nothing until its configuration is
fetched. PageStart measures each banner's height only when its onRender
callback changes identity, which happens when PageStart re-renders - not
when this banner later becomes visible. Its height therefore stayed 0, so
the next banner stacked on top of it and the two overlapped.

Re-report the height whenever the banner's visibility or dates change, so
PageStart can place the following banner below it.

sdumitriu commented on modules/demo-banner/src/main/resources/SLING-INF/content/Extensions/SurveyPageStart/WarningBanner.json:
Not a problem for this PR, but these should be named `DemoWarningBanner` or `DemoBanner`, not the generic `WarningBanner`. Warning about what?

sdumitriu commented on modules/demo-banner/pom.xml:
```diff
@@ -45,6 +45,7 @@
             <Sling-Initial-Content>
               SLING-INF/content/Extensions/PageStart/;path:=/Extensions/PageStart/;overwriteProperties:=true;uninstall:=true,
               SLING-INF/content/Extensions/LoginPageStart/;path:=/Extensions/LoginPageStart/;overwriteProperties:=true;uninstall:=true,
+              SLING-INF/content/Extensions/SurveyPageStart/;path:=/Extensions/SurveyPageStart/;overwriteProperties:=true;uninstall:=true
```
Having to copy the same thing 3 times is too much. Other than specific endpoints for specific pages, there should also be a `GlobalPageStart` extension point that is always used by all our screens, including error pages, expired, unsubscribe...

sdumitriu replied:
https://phenotips.atlassian.net/browse/CARDS-2921

sdumitriu commented on modules/patient-portal/src/main/resources/SLING-INF/content/apps/cards/ExtensionPoints/PatientPortalPageStart.json:
Since extension point nodes are global, does it make sense to name this `PatientPortalPageStart` or `PatientSurveyPageStart`?

marta- replied:
PatientPortalPageStart

sdumitriu reviewed (commented):
Looks good, two comments for the future and one question for now:

sdumitriu committed:
CARDS-1741: Add a "PageStart" extension point to the patient UI
Rename the extension point node name to something more specific to avoid future conflicts
Also rename the existing Footer extension point
